From 4bbdf750148e9e163137e51575cf1c78145fffc1 Mon Sep 17 00:00:00 2001
From: Colin Patrick Mccabe <cmccabe@cloudera.com>
Date: Wed, 8 Oct 2014 17:55:46 -0700
Subject: [PATCH 0631/1023] HDFS-7202. Should be able to omit package name of
 SpanReceiver on "hadoop trace -add" (iwasakims
 via cmccabe)

(cherry picked from commit d996235285e5047f731e3d3fc4c6e6214caa10aa)
(cherry picked from commit f1feaae1dd036ab77fe50d392e6fbf560541020e)
---
 .../apache/hadoop/tracing/SpanReceiverHost.java    |    8 ++++----
 .../org/apache/hadoop/tracing/TestTraceAdmin.java  |    8 ++++++++
 2 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/tracing/SpanReceiverHost.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/tracing/SpanReceiverHost.java
index 82f099e..13c92ac 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/tracing/SpanReceiverHost.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/tracing/SpanReceiverHost.java
@@ -137,10 +137,6 @@ public synchronized void loadSpanReceivers(Configuration conf) {
           getUniqueLocalTraceFileName());
     }
     for (String className : receiverNames) {
-      className = className.trim();
-      if (!className.contains(".")) {
-        className = "org.htrace.impl." + className;
-      }
       try {
         SpanReceiver rcvr = loadInstance(className, EMPTY);
         Trace.addReceiver(rcvr);
@@ -154,6 +150,10 @@ public synchronized void loadSpanReceivers(Configuration conf) {
 
   private synchronized SpanReceiver loadInstance(String className,
       List<ConfigurationPair> extraConfig) throws IOException {
+    className = className.trim();
+    if (!className.contains(".")) {
+      className = "org.htrace.impl." + className;
+    }
     Class<?> implClass = null;
     SpanReceiver impl;
     try {
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/tracing/TestTraceAdmin.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/tracing/TestTraceAdmin.java
index 74f51c0..76cdb96 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/tracing/TestTraceAdmin.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/tracing/TestTraceAdmin.java
@@ -83,6 +83,14 @@ public void testCreateAndDestroySpanReceiver() throws Exception {
               getHostPortForNN(cluster)));
       Assert.assertEquals("ret:0, [no span receivers found]" + NEWLINE,
           runTraceCommand(trace, "-list", "-host", getHostPortForNN(cluster)));
+      Assert.assertEquals("ret:0, Added trace span receiver 2 with " +
+          "configuration local-file-span-receiver.path = " + tracePath + "\n",
+          runTraceCommand(trace, "-add", "-host", getHostPortForNN(cluster),
+              "-class", "LocalFileSpanReceiver",
+              "-Clocal-file-span-receiver.path=" + tracePath));
+      Assert.assertEquals("ret:0, Removed trace span receiver 2\n",
+          runTraceCommand(trace, "-remove", "2", "-host",
+              getHostPortForNN(cluster)));
     } finally {
       cluster.shutdown();
       tempDir.close();
-- 
1.7.9.5

